README.md
## PR-DC MQTT Node.js

Example of MQTT Broker and Client implemented in Node.js.

For MQTT Web Client checkout https://github.com/PR-DC/PRDC_MQTTWebClient

## Requirements
[Node.js](https://nodejs.org/en/)<br>

This code is tested with
**Node.js 16.1.0**

## Installation

On Windows from `scripts\win10\` first run `installNodejs.sh` and then `installApp.cmd`, for both Broker and Client.

On Linux from `scripts\linux\` first run `installNVM.sh` then `installNodejs.sh` and finally `runApp.sh`, for both Broker and Client.

## Results

After running both Broker and Client you should get two windows similar to the following.

<p align="center">
  <img src="https://pr-dc.com/web/img/github/PRDC_MQTT_NodeJS.jpg" width="800">
</p>

## License
Copyright (C) 2021 PR-DC <info@pr-dc.com>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as 
published by the Free Software Foundation, either version 3 of the 
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.


CHANGES.txt
--------------------
PRDC_MQTT_NodeJS
Example of MQTT Node.js Broker and Client by PR-DC
--------------------

Release 1.0.0, 12.12.2021.

- Initial release

--------------------

PRDC_MQTT_CLIENT/PRDC_MQTT_CLIENT.js
/**
 * PRDC_MQTT_CLIENT - PR-DC MQTT NodeJS Client
 * Author: Milos Petrasinovic <mpetrasinovic@pr-dc.com>
 * PR-DC, Republic of Serbia
 * info@pr-dc.com
 * 
 * --------------------
 * Copyright (C) 2021 PR-DC <info@pr-dc.com>
 *
 * This file is part of PRDC_MQTT_NodeJS.
 *
 * PRDC_MQTT_NodeJS is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as 
 * published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version.
 * 
 * PRDC_MQTT_NodeJS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with PRDC_MQTT_NodeJS.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

// Configuration
// --------------------
const broker_host = 'localhost'; // enter broker host address
const broker_port = 1883;
var username = 'publisher'; // [-] MQTT Client username
var password = 'some_pw_for_publisher'; // [-] MQTT Client password
var topic = 'some_topic'; // [-] MQTT topic
var qos = 0; // [-] MQTT quality of service 
var retain = true; // [-] retain MQTT message
var connection_timeout = 3000; // [s] time for connection timeout
var keep_alive = 60; // [s] time for keep alive signal
var client_id = 'PRDC_MQTT_Client' + parseInt(Math.random() * 100000);
var publish_dt = 3000;
var i = 0;
var pause = false;

// Modules 
// --------------------
var MQTT = require('mqtt');

// Connect to broker 
// --------------------
var mqttClient = MQTT.connect({
  host: broker_host,
  port: broker_port,
  clientId: client_id,
  username: username,
  password: password,
  connectTimeout: connection_timeout,
  keepalive: keep_alive
});

// On keypress
const readline = require('readline');
readline.emitKeypressEvents(process.stdin);
process.stdin.setRawMode(true);
process.stdin.on('keypress', (str, key) => {
  if (key.ctrl && key.name === 'p') {
    if(pause) {
      pause = false;
      console.log('Execution resumed...');
    } else {
      pause = true;
      console.log('Execution paused, press [CTRL+P] to resume...');
    }
  }
});

// Connect and subscribe
mqttClient.on('connect', function(message) {
  showMessage('Connected to ' + broker_host + ':' + broker_port);
  setInterval(publishMessage, publish_dt);
  
  mqttClient.subscribe(topic, function(e) {
    if(!e) {
     showMessage('Subscribed to topic: ' +topic);
    }
  });
});


// Message arrived
mqttClient.on('message', function (topic, message, packet) {
  showMessage('Client published: ' + message.toString() + 
    ' | Topic: ' + topic);
})

// Publish new message
function publishMessage() {
  if(!pause) {
    var msg = 'Sending packet ' + i;
    i += 1;
    
    mqttClient.publish(topic, msg);
    showMessage('Publishing: ' + msg + ' | Topic: ' + topic);
    console.log('Press [CTRL+P] to pause execution...');
  }
}

// Functions
// --------------------

// getTimestamp() function
// Returns current timestamp in format HH:mm:SS.ms
// --------------------
function getTimestamp() {
  var date = new Date();
  var pad = function(num, size) { 
    return ('000' + num).slice(size * -1); 
  };
  var time = parseFloat(date.getTime()/1000).toFixed(3);
  var hours = date.getHours();
  var minutes = Math.floor(time / 60) % 60;
  var seconds = Math.floor(time - minutes * 60);
  var milliseconds = time.slice(-3);

  return pad(hours, 2) + ':' + pad(minutes, 2) + ':' + 
    pad(seconds, 2) + '.' + pad(milliseconds, 3);
}

// showMessage() function
// Show message
// --------------------
function showMessage(msg) {
  console.log('\x1b[36m['+getTimestamp()+']\x1b[0m ' + msg);
}


PRDC_MQTT_CLIENT/scripts/linux/runApp.sh
#!/bin/sh
# runApp.sh
# Batch file for running of NodeJS app
# Author: Milos Petrasinovic <mpetrasinovic@pr-dc.com>
# PR-DC, Republic of Serbia
# info@pr-dc.com
# --- INPUTS ---
#  nodejsVersion - version of node.js
# --------------------
#
# Copyright (C) 2021 PR-DC <info@pr-dc.com>
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as 
# published by the Free Software Foundation, either version 3 of the 
# License, or (at your option) any later version.
#  
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#  
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --------------------

nodejsVersion="16.1.0"

# Execute commands
. ~/.nvm/nvm.sh
. ~/.profile
. ~/.bashrc
cd ../../ 
nvm use $nodejsVersion
npm start
# --------------------


PRDC_MQTT_CLIENT/scripts/linux/installNVM.sh
#!/bin/sh
# installNVM.sh
# Batch file for installation of NVM for NodeJS app
# Author: Milos Petrasinovic <mpetrasinovic@pr-dc.com>
# PR-DC, Republic of Serbia
# info@pr-dc.com
# --------------------
#
# Copyright (C) 2021 PR-DC <info@pr-dc.com>
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as 
# published by the Free Software Foundation, either version 3 of the 
# License, or (at your option) any later version.
#  
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#  
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --------------------

# Execute commands
sudo apt install curl
curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.38.0/install.sh | bash
# --------------------


PRDC_MQTT_BROKER/scripts/linux/installNodejs.sh
#!/bin/sh
# installNodejs.sh
# Batch file for installation of a certain version of node.js
# Author: Milos Petrasinovic <mpetrasinovic@pr-dc.com>
# PR-DC, Republic of Serbia
# info@pr-dc.com
# --- INPUTS ---
#  nodejsVersion - version of node.js
# --------------------
#
# Copyright (C) 2021 PR-DC <info@pr-dc.com>
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Lesser General Public License as 
# published by the Free Software Foundation, either version 3 of the 
# License, or (at your option) any later version.
#  
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Lesser General Public License for more details.
#  
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# --------------------

nodejsVersion="16.1.0"

# Execute commands
. ~/.nvm/nvm.sh
. ~/.profile
. ~/.bashrc
nvm install $nodejsVersion
# --------------------


PRDC_MQTT_BROKER/PRDC_MQTT_BROKER.js
/**
 * PRDC_MQTT_BROKER - PR-DC MQTT NodeJS Broker
 * Author: Milos Petrasinovic <mpetrasinovic@pr-dc.com>
 * PR-DC, Republic of Serbia
 * info@pr-dc.com
 * 
 * --------------------
 * Copyright (C) 2021 PR-DC <info@pr-dc.com>
 *
 * This file is part of PRDC_MQTT_NodeJS.
 *
 * PRDC_MQTT_NodeJS is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as 
 * published by the Free Software Foundation, either version 3 of the 
 * License, or (at your option) any later version.
 * 
 * PRDC_MQTT_NodeJS is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with PRDC_MQTT_NodeJS.  If not, see <https://www.gnu.org/licenses/>.
 *
 */

// Configuration
// --------------------
const broker_id = 'PR-DC_MQTT_BROKER'
const port = 1883;
const http_port = 8888;
const client_username = 'prdc';
const client_password = 'some_password';
const publish_username = 'publisher';
const publish_password = 'some_pw_for_publisher';

// Get IP adresses
// --------------------
const { networkInterfaces } = require('os');
const nets = networkInterfaces();
const results = {}; // Or just '{}', an empty object
const publicIp = require('public-ip');
var local_ip = '0.0.0.0';
var public_ip = '0.0.0.0';

(async () => {
  public_ip = await publicIp.v4()
  showMessage('Server public IP adress: ' + public_ip);
})();

for(const name of Object.keys(nets)) {
  for(const net of nets[name]) {
    // Skip over non-IPv4 and internal (i.e. 127.0.0.1) addresses
    if(net.family === 'IPv4' && !net.internal) {
      if(!results[name]) {
        results[name] = [];
      }
      results[name].push(net.address);
      if(name === 'Ethernet' || name === 'en0') {
        local_ip = net.address;
      }
    }
  }
}
showMessage('Server local IP adress: ' + local_ip);

// Logs
const fs = require('fs');

// Modules 
// --------------------
const MQTT = require('aedes')({
  id: broker_id,
  authenticate: function (client, username, password, callback) {
    if((username === client_username && 
        password.toString() === client_password) ||
        username === publish_username && 
        password.toString() === publish_password) {
      client.username = username;
      client.password = password;
      callback(null, true);
    } else {
      var err = new Error('Wrong username or password.');
      err.returnCode = 4;
      callback(err, false);
    }
  },
  authorizePublish: function (client, packet, callback) {
    if(client.username !== publish_username) {
      return callback(new Error('Unauthorized publish.'));
    }
    client.publish_topics = [];
    client.publish_log = true;
    callback(null);
  },
  authorizeSubscribe: function (client, sub, callback) {
    callback(null, sub);
  }
});

const ws = require('websocket-stream');
const net = require('net');
const http = require('http');

// Create servers
// --------------------
const server = net.createServer(MQTT.handle);
const httpServer = http.createServer();
ws.createServer({server: httpServer}, MQTT.handle);

// Run servers
// --------------------
server.listen(port, function() {
   showMessage('Server started and listening on port ' + port);
  
  MQTT.publish({ 
    topic: 'PR-DC/hello', 
    payload: 'I am PR-DC MQTT broker ' + MQTT.id 
  });
});

httpServer.listen(http_port, function() {
   showMessage('WebSocket server listening on port ' + http_port);
  
  MQTT.publish({ 
    topic: 'PR-DC/hello', 
    payload: 'I am PR-DC WebSocket MQTT broker ' + MQTT.id 
  });
});

/* MQTT.subscribe('PR-DC/hello', function(packet, cb) {
   showMessage('Published' + packet.payload.toString());

  MQTT.publish({
    topic: 'PR-DC/hello',
    payload: 'Thank you!'
  });
});
 */

// Server events
// --------------------
// Client connects
MQTT.on('client', function(client) {
  showMessage('Client \x1b[33m' + (client ? client.id : client) + 
    '\x1b[0m connected to broker ' + MQTT.id)
})

// Client disconnects
MQTT.on('clientDisconnect', function(client) {
  showMessage('Client \x1b[31m' + (client ? client.id : client) + 
    '\x1b[0m disconnected from broker ' + MQTT.id)
})

// On subscribe
MQTT.on('subscribe', function(subscriptions, client) {
  showMessage('Client \x1b[32m' + (client ? client.id : client) +
    '\x1b[0m subscribed to topics: ' + 
    subscriptions.map(s => s.topic).join('\n'))
})

// On unsubscribe
MQTT.on('unsubscribe', function(subscriptions, client) {
   showMessage('Client \x1b[32m' + (client ? client.id : client) +
    '\x1b[0m unsubscribed from topics: ' + subscriptions.join('\n'))
})

// Message is published
MQTT.on('publish', async function(packet, client) {
  showMessage('Client \x1b[31m' + (client ? client.id : MQTT.id) + 
    '\x1b[0m has published: ' + packet.payload.toString() + 
    ' | Topic: ' + packet.topic)
  if(client && client.publish_log) {
    if(packet.topic in client.publish_topics.includes) {
      var fid = client.publish_topics[packet.topic];
    } else {
      var fid = fs.createWriteStream('logs/' + client.id + '_' + packet.topic + '.log', {
        flags: 'a'
      })
      client.publish_topics[packet.topic] = fid;
    }
    fid.write(packet.payload);
  }
})

// Functions
// --------------------

// getTimestamp() function
// Returns current timestamp in format HH:mm:SS.ms
// --------------------
function getTimestamp() {
  var date = new Date();
  var pad = function(num, size) { 
    return ('000' + num).slice(size * -1); 
  };
  var time = parseFloat(date.getTime()/1000).toFixed(3);
  var hours = date.getHours();
  var minutes = Math.floor(time / 60) % 60;
  var seconds = Math.floor(time - minutes * 60);
  var milliseconds = time.slice(-3);

  return pad(hours, 2) + ':' + pad(minutes, 2) + ':' + 
    pad(seconds, 2) + '.' + pad(milliseconds, 3);
}

// showMessage() function
// Show message
// --------------------
function showMessage(msg) {
  console.log('\x1b[36m['+getTimestamp()+']\x1b[0m ' + msg);
}
